File: apps/02-personal-md-files/backend/nodejs/src/controllers/chatController.js
import { llmService } from '../services/llmService.js';
import { markdownService } from '../services/markdownService.js';


function generatePrompt(message, markdownFiles) {
  return `You are a helpful team lead. You are given a question and  information about the team members. Please answer the question based on the information provided.
    You provide answer only based on the information provided in the context.
    You do not generate any information outside of the context.
    If you do not know the answer, just say "I do not know".
    <question>${message}</question>\n\n
    <context>${markdownFiles.map(file => '<'+file.filename+'>'+file.content+'</'+file.filename+'>').join('\n\n')}</context>`
};

export const chatController = {
  async chat(req, res) {
    try {
      const { message, provider = 'ollama' } = req.body;

      if (!message) {
        return res.status(400).json({ error: 'Message is required' });
      }

      const markdownFiles = await markdownService.readAllMarkdownFiles();
      const prompt = generatePrompt(message, markdownFiles);
      const response = await llmService.generateResponse(prompt, provider);
      res.json({ response });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },


  async providers(req, res) {
    try {
      const providers = llmService.getAvailableProviders();
      res.json({ providers });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  async switchProvider(req, res) {
    try {
      const { provider } = req.body;

      if (!provider) {
        return res.status(400).json({ error: 'Provider is required' });
      }

      if (!llmService.isProviderAvailable(provider)) {
        return res.status(400).json({ error: 'Provider not available' });
      }

      res.json({ message: 'Provider switched successfully' });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  }
}; 

File: apps/02-personal-md-files/backend/python/src/controllers/chat_controller.py
from typing import Dict, Any
from ..services.llm_service import LLMService
from ..services.markdown_service import MarkdownService

class ChatController:
    """Controller for chat functionality"""
    
    def __init__(self, llm_service:LLMService, markdown_service:MarkdownService):
        self.llm_service = llm_service
        self.markdown_service = markdown_service
    
    async def chat(self, data: Dict[str, Any]) -> Dict[str, str]:
        """Handle chat message requests"""
        if "message" not in data:
            raise Exception("Message is required")
        
        message = data['message']
        provider = data.get('provider', 'ollama')
        
        try:
            # Get all markdown files content
            markdown_files = self.markdown_service.read_all_markdown_files()

            # Generate prompt with user message and context
            prompt = self.generate_prompt(message, markdown_files)

            # Send the prompt to the LLM service
            response = await self.llm_service.send_message(prompt, provider)

            return {'response': response}
        except Exception as e:
            raise Exception(f"Failed to generate response: {str(e)}")
    


    def generate_prompt(self, message, markdown_files) -> str:
        static_text = """You are a helpful team lead. You are given a question and  information about the team members. Please answer the question based on the information provided.
      You provide answer only based on the information provided in the context.
      You do not generate any information outside of the context.
      If you do not know the answer, just say "I do not know"."""
        
        question_part = f"<question>{message}</question>"
        
        context_part = '<context>'
        for file in markdown_files:
            context_part += f"<{file['filename']}>"
            context_part += file['content']
            context_part += f"</{file['filename']}>"
        context_part += '</context>'
        
        return f"{static_text}\n\n{question_part}\n\n{context_part}"

    def providers(self) -> Dict[str, str]:
        try:
            providers = self.llm_service.get_available_providers()
            return {'providers': providers}
        except Exception as e:
            raise Exception(f"Failed to get providers: {str(e)}")

    def switch_provider(self, data: Dict[str, Any]) -> Dict[str, str]:
        """Switch the current LLM provider"""
        if "provider" not in data:
            raise Exception("Provider is required")
        
        provider = data['provider']

        if not self.llm_service.is_provider_available(provider):
            raise Exception("Provider not available")

        return {'message': 'Provider switched successfully'}